### PR TITLE
[SPARK-50341][SS][PYTHON] Use UDS for JVM and Python worker communication

### PR DESCRIPTION
**Pull request by bogao007**
### What changes were proposed in this pull request?

Use UDS for JVM and Python worker communication. This is a followup item from a previous PR comment https://github.com/apache/spark/pull/47133#discussion_r1680043141

### Why are the changes needed?

UDS is recommended for same-host processes communication.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasStateServer.scala [hunk 1]
```diff
@@ -758,3 +760,12 @@ case class MapStateInfo(
     keySerializer: ExpressionEncoder.Serializer[Row],
     valueDeserializer: ExpressionEncoder.Deserializer[Row],
     valueSerializer: ExpressionEncoder.Serializer[Row])
+
+object TransformWithStateInPandasStateServer {
+  @volatile private var id = 0
```
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasStateServer.scala:765]: IIUC, `TransformWithStateInPandasPythonRunner` is instantiated per executor so they are in different JVMs, and `TransformWithStateInPandasStateServer` is initialized from each `TransformWithStateInPandasPythonRunner` once. What could be the scenario for handling multiple threads here?

**Review by jingz-db — APPROVED** [review]
Left a small question otherwise looks good. Thanks for making the change!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasStateServer.scala (hunk 1 above)
Comment (bogao007) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasStateServer.scala:765]: There's no much difference comparing to TCP socket. For UDS, we create socket files for each server thread connection, and the files are named with different `serverId` which got updated incrementally. At the end we do cleanup to delete these socket files.

**Comment by HyukjinKwon** [comment]
> UDS is recommended for same-host processes communication

Should we better change all usage in Python worker communication instead of this alone?

File: sql/core/pom.xml [hunk 2]
```diff
@@ -252,6 +252,11 @@
       <artifactId>htmlunit3-driver</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (LuciferYang) [inline sql/core/pom.xml:255]: should add this to dependency management first

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 3]
```diff
@@ -21,6 +21,11 @@ arrow-memory-core/18.0.0//arrow-memory-core-18.0.0.jar
 arrow-memory-netty-buffer-patch/18.0.0//arrow-memory-netty-buffer-patch-18.0.0.jar
 arrow-memory-netty/18.0.0//arrow-memory-netty-18.0.0.jar
 arrow-vector/18.0.0//arrow-vector-18.0.0.jar
+asm-analysis/9.2//asm-analysis-9.2.jar
+asm-commons/9.2//asm-commons-9.2.jar
+asm-tree/9.2//asm-tree-9.2.jar
+asm-util/9.2//asm-util-9.2.jar
+asm/9.2//asm-9.2.jar
```
Comment (LuciferYang) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:28]: Spark  master branch is currently using `asm` version 9.7.1, and I think we should unify to version 9.7.1.

**Comment by LuciferYang** [comment]
Is this a Unix-only feature? For Windows users, will it fail or are there workarounds available?

**Comment by LuciferYang** [comment]
> > UDS is recommended for same-host processes communication
> 
> Should we better change all usage in Python worker communication instead of this alone?

+1

And since I'm not a professional Python developer, please allow me to ask a beginner's question: Is it a consensus that `UDS is recommended for same-host process communication`? Are there any relevant test data to support this?

File: sql/core/pom.xml (hunk 2 above)
Comment (yaooqinn) [reply sql/core/pom.xml:255]: For non-ASF ones, we will include their licenses in our licenses-binary directory and LICENSE-binary file, and also update the NOTICE-binary document for any applicable notices.

File: python/pyspark/worker.py [hunk 4]
```diff
@@ -1449,7 +1449,7 @@ def mapper(_, it):
 def read_udfs(pickleSer, infile, eval_type):
     runner_conf = {}
 
-    state_server_port = None
+    state_server_id = None
```
Comment (allisonwang-db) [inline python/pyspark/worker.py:1452]: Why do we need this change?

**Review by allisonwang-db — COMMENTED** [review]
Can we make the PR title more explicit? Add the `FOLLOWUP` tag and also mentioning this is for streaming Python worker only?

**Comment by bogao007** [comment]
> > > UDS is recommended for same-host processes communication
> > 
> > 
> > Should we better change all usage in Python worker communication instead of this alone?
> 
> +1
> 
> And since I'm not a Python developer, please allow me to ask a beginner's question: Is it a consensus that `UDS is recommended for same-host process communication`? Are there any relevant test data to support this?

@LuciferYang Thanks for bringing this up! Unfortunately I don't have any test data to back this statement. Though it might have a slightly better performance, it may also bring some downsides compared to TCP socket (e.g. Windows compatibility). I'll drop this PR for now given not enough evidence to do the UDS migration, thanks!

**Pull request closed without merging**